discovery_engine_core/bindings/src/tracing.rs

```diff
@@ -14,37 +14,105 @@
 
 //! Setup tracing on different platform.
 
-use std::sync::Once;
+use std::{env::VarError, error::Error, fs::OpenOptions, io, path::Path, sync::Once};
 
-use tracing_subscriber::{filter::LevelFilter, util::SubscriberInitExt};
+use tracing::metadata::LevelFilter;
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};
 
 static INIT_TRACING: Once = Once::new();
 
-pub(crate) fn init_tracing() {
+pub(crate) fn init_tracing(log_file: Option<&Path>) {
     INIT_TRACING.call_once(|| {
-        init_tracing_once();
+        init_tracing_once(log_file);
+        init_panic_logging();
     });
 }
 
-fn init_tracing_once() {
-    let subscriber = tracing_subscriber::fmt()
-        .with_max_level(LevelFilter::INFO)
-        .finish();
+fn init_tracing_once(log_file: Option<&Path>) {
+    let mut delayed_errors = Vec::new();
 
+    let stdout_log = tracing_subscriber::fmt::layer();
+
+    #[allow(clippy::needless_late_init)]
+    let android_logging;
     cfg_if::cfg_if! {
         if #[cfg(target_os = "android")] {
-            use tracing_subscriber::layer::SubscriberExt;
-            let layer = tracing_android::layer("xayn_discovery_engine").ok();
-            subscriber.with(layer).init();
+            android_logging = tracing_android::layer("xayn_discovery_engine").ok()
         } else {
-            subscriber.init();
+            // workaround to not have to specify a alternative type per hand
+            android_logging = false.then(tracing_subscriber::fmt::layer)
         }
+    };
+
+    let file_log = log_file
+        .map(|log_file| -> io::Result<_> {
+            let writer = OpenOptions::new()
+                .write(true)
+                .truncate(true)
+                .create(true)
+                .open(log_file)?;
+
+            Ok(tracing_subscriber::fmt::layer().with_writer(writer).json())
+        })
+        .transpose()
+        .map_err(|err| {
+            delayed_errors.push(Box::new(err) as _);
+        })
+        .ok();
+
+    let filter = build_env_filter(&mut delayed_errors);
+
+    tracing_subscriber::registry()
+        .with(stdout_log)
+        .with(android_logging)
+        .with(file_log)
+        .with(filter)
+        .init();
+
+    for error in delayed_errors {
+        tracing::error!(error=%error, "logging setup failed");
```

Comment:
🎨 the local variable can be used directly without a field name in case of same names
```suggestion
        tracing::error!(%error, "logging setup failed");
```